WINC-805: Add support for Windows Server 2022 in AWS

jrvaldes opened the pull request:
This PR updates the AWS's EC2 AMI to use Windows Server 2022
Base image without Docker container runtime, as the container runtime
(containerd) is installed by WMCO.

Leverage the environment variable introduced in the release job configuration to use Windows Server 2019 as machine and container images for the `aws-e2e-operator` test
- https://github.com/openshift/release/pull/29791

mansikulkarni96 commented:
/test aws-e2e-operator

jrvaldes commented:
/test aws-e2e-operator

mansikulkarni96 commented:
/test aws-e2e-upgrade

saifshaikh48 commented on docs/machineset-aws.md:
```diff
@@ -1,9 +1,14 @@
 # Creating an AWS Windows MachineSet
 
-_\<windows_container_ami\>_ can be used as it is without any modification.
-You must use Windows Server 2019 with a version 10.0.17763.1457 or earlier to work
-around Windows containers behind a Kubernetes load balancer becoming unreachable
-[issue](https://github.com/microsoft/Windows-Containers/issues/78).
```
is this guidance no longer relevant?

selansen replied:
I think it should be like 'you can use 2019 or 2022` as a suggestion

jrvaldes replied:
I don't think we should duplicate the [Supported Windows Server versions](https://github.com/openshift/windows-machine-config-operator/blob/master/docs/wmco-prerequisites.md#supported-windows-server-versions). I will reword this to include a reference.

selansen commented:
Can you pls add a few lines in the description section?

mansikulkarni96 commented:
Don't we need to update the image deployed for networking test [here](https://github.com/openshift/windows-machine-config-operator/blob/437607fefb30ca40547f08b2ad24dab15ac9d25e/test/e2e/network_test.go#L458)

jrvaldes commented:
> Don't we need to update the image deployed for networking test [here](https://github.com/openshift/windows-machine-config-operator/blob/437607fefb30ca40547f08b2ad24dab15ac9d25e/test/e2e/network_test.go#L458)

Yes

jrvaldes commented on docs/machineset-aws.md:
```diff
@@ -1,9 +1,14 @@
-You must use Windows Server 2019 with a version 10.0.17763.1457 or earlier to work
-around Windows containers behind a Kubernetes load balancer becoming unreachable
-[issue](https://github.com/microsoft/Windows-Containers/issues/78).
+_\<windows_ami_id\>_ should be replaced with the output of:
```
Makes sense, will move it after the region param.

jrvaldes commented:
/test aws-e2e-operator

jrvaldes commented:
/test aws-e2e-operator

jrvaldes commented:
/test aws-e2e-operator

jrvaldes commented:
/test aws-e2e-operator

jrvaldes commented:
/test aws-e2e-operator

jrvaldes commented:
/test aws-e2e-upgrade

jrvaldes commented:
/test aws-e2e-upgrade

jrvaldes commented:
/test aws-e2e-operator

Build src image [failed](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_windows-machine-config-operator/1091/pull-ci-openshift-windows-machine-config-operator-master-aws-e2e-upgrade/1536721341460254720)

jrvaldes commented:
/test aws-e2e-operator

Retrying, build src image [failed](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_windows-machine-config-operator/1091/pull-ci-openshift-windows-machine-config-operator-master-aws-e2e-operator/1536725835644407808)

jrvaldes commented:
/test aws-e2e-operator

jrvaldes commented:
/test aws-e2e-operator

jrvaldes commented:
/test aws-e2e-upgrade

jrvaldes commented:
/test aws-e2e-operato

jrvaldes commented:
/test aws-e2e-operator

jrvaldes commented:
/test aws-e2e-upgrade

jrvaldes commented:
/test azure-e2e-operator

jrvaldes commented:
/test aws-e2e-operator

jrvaldes commented:
/test aws-e2e-upgrade

jrvaldes commented:
/test aws-e2e-upgrade

jrvaldes commented:
/test aws-e2e-operator

jrvaldes commented:
/test aws-e2e-upgrade

jrvaldes commented:
/test aws-e2e-operator

jrvaldes commented:
/test aws-e2e-upgrade

jrvaldes commented:
/test aws-e2e-operator

jrvaldes commented:
/test lint

jrvaldes commented:
/test aws-e2e-operator

jrvaldes commented:
/test aws-e2e-operator

saifshaikh48 commented on test/e2e/validation_test.go:
```diff
@@ -309,9 +309,10 @@ func (tc *testContext) sshSetup() error {
 	powershellDefaultCommand := command
-	if tc.hasCustomVXLAN {
+	if tc.CloudProvider.GetType() == config.VSpherePlatformType ||
+		tc.CloudProvider.GetType() == config.AWSPlatformType {
```
I have a feeling this logic won't suffice if we are testing both 2019 and 2022 using AWS jobs. One suggestion is to replace `tc.hasCustomVXLAN` with a flag `isWindowsServer2022` in the [`testContext` struct](https://github.com/openshift/windows-machine-config-operator/blob/master/test/e2e/main_test.go#L64), setting it at the time of [machineset creation](https://github.com/openshift/windows-machine-config-operator/blob/2dfa5ee1d5b6223c11c6c27353b7429e99253b23/test/e2e/create_test.go#L262) in the e2e test suite. Might require some parsing of the machine set spec to see the OS version of the image being used.

jrvaldes commented:
/hold cancel

https://github.com/openshift/windows-machine-config-operator/pull/1101 has merged.

jrvaldes commented:
/test aws-e2e-operator

jrvaldes commented:
/test aws-e2e-upgrade

jrvaldes commented:
/test aws-e2e-ccm-install

jrvaldes commented:
/hold 

needs rebase with:
- https://github.com/openshift/windows-machine-config-operator/pull/1093

jrvaldes commented:
/test aws-e2e-upgrade

jrvaldes commented:
/test aws-e2e-ccm-install

jrvaldes commented:
/test lint

jrvaldes commented:
/test unit

jrvaldes commented on test/e2e/network_test.go:
```diff
@@ -440,24 +440,19 @@ func (tc *testContext) getPodIP(selector metav1.LabelSelector) (string, error) {
-		// If we're on vSphere we will use 2022
-		windowsServerImage = "mcr.microsoft.com/powershell:lts-nanoserver-ltsc2022"
-	} else if tc.CloudProvider.GetType() == config.NonePlatformType {
+	if tc.CloudProvider.GetType() == config.NonePlatformType {
```
With the latest rebase from https://github.com/openshift/windows-machine-config-operator/pull/1109 , the multiple if statement is no longer present.

jrvaldes commented:
/hold cancel 

https://github.com/openshift/windows-machine-config-operator/pull/1093 merged

sebsoto commented:
/hold
active discussion on
https://github.com/openshift/release/pull/29791/commits/be0539ac05b6b22d92856344ba4d0a4532b6fb41#r908822090

jrvaldes commented on test/e2e/validation_test.go:
```diff
@@ -308,11 +308,8 @@ func (tc *testContext) sshSetup() error {
-		powershellDefaultCommand = strings.ReplaceAll(command, "\\\"", "\"")
-	}
+	// Do not scape the double-quotes when default shell is Powershell
+	powershellDefaultCommand := strings.ReplaceAll(command, "\\\"", "\"")
```
As WS2019 will be used in the e2e test suite along with the env var `WINDOWS_SERVER_2019_ENABLED`, this logic should be adjusted to capture that.

saifshaikh48 replied:
+1 to adding the `WINDOWS_SERVER_VERSION` env var information to testContext struct so eventually this can look like
```suggestion
	powershellDefaultCommand := command
	if tc.windowsServerVersion == "2022" {
		powershellDefaultCommand = strings.ReplaceAll(command, "\\\"", "\"")
	}
```

jrvaldes replied:
For further references, the double quotes should be adjusted only when the default shell is Powershell.

| Default shell    | OS version    | Need \\" replacement in command?|
| --------------------|-------------------|--------------------------- |
| Windows CMD      | 2019          | No |
| Powershell       | 2019          | No |
| Windows CMD      | 2022          | No |
| Powershell       | 2022          | Yes |

Thanks @saifshaikh48 for sharing this table.

jrvaldes commented on test/e2e/network_test.go:
```diff
@@ -440,17 +440,13 @@ func (tc *testContext) getPodIP(selector metav1.LabelSelector) (string, error) {
 	return podList.Items[0].Status.PodIP, nil
 }
 
```
You are right, they can be squashed all together.

sebsoto commented on test/e2e/network_test.go:
```diff
@@ -440,19 +440,22 @@ func (tc *testContext) getPodIP(selector metav1.LabelSelector) (string, error) {
-	if tc.CloudProvider.GetType() == config.AWSPlatformType {
-		// On AWS we are currently testing 2019
+	// Check the Windows Server version environment variable to set the correct container image
+	windowsServerVersion := os.Getenv("WINDOWS_SERVER_VERSION")
```
doing this check at the creation of `testContext`, and making `windowsServerVersion` part of the testContext struct will allow us to avoid having to make this syscall multiple times

jrvaldes replied:
Addressed in https://github.com/openshift/windows-machine-config-operator/commit/c0b8dc12ef4652bca2551d4af0fa2d116b9b9984

sebsoto commented on test/e2e/providers/aws/aws.go:
```diff
@@ -114,18 +114,33 @@ func (a *awsProvider) getInfraID() (string, error) {
+	// The Windows Server version can be provided as an environment variable. For example, from the CI job definition
+	// in the release step. The selected Windows Server version must be compatible with Windows container image.
+	// See https://docs.microsoft.com/en-us/virtualization/windowscontainers/deploy-containers/base-image-lifecycle
+	windowsServerVersion := os.Getenv("WINDOWS_SERVER_VERSION")
```
please make this check within `SetupAWSCloudProvider` and make it part of the awsProvider struct, for the same reasons as above

jrvaldes replied:
In contrast with above case, the cardinality of the  `SetupAWSCloudProvider` func with the proposed `getWindowsAMIFilter` func is one to one. So that, the syscall to read the env var (`os.Getenv()`) is executed just one time in the execution flow.  IMO, the `windowsServerVersion` field is specific to the e2e test and should not be part of the `awsProvider` struct.

jrvaldes commented on test/e2e/validation_test.go:
```diff
@@ -308,12 +308,8 @@ func (tc *testContext) sshSetup() error {
-	if tc.hasCustomVXLAN {
-		powershellDefaultCommand = strings.ReplaceAll(command, "\\\"", "\"")
-	}
+	// Do not scape the double-quotes when default shell is Powershell
```
Fixed

jrvaldes commented:
/test aws-e2e-operator

saifshaikh48 commented on test/e2e/providers/aws/aws.go:
```diff
@@ -114,18 +114,33 @@ func (a *awsProvider) getInfraID() (string, error) {
+	windowsServerVersion := os.Getenv("WINDOWS_SERVER_VERSION")
+	switch windowsServerVersion {
+	case "2019":
+		return "Windows_Server-2019-English-Full-ContainersLatest-????.??.??"
```
I guess we could continue using with `Containers` so we can ensure having docker installed doesn’t mess anything up.

jrvaldes replied:
Correct, I'm in the same boat.

jrvaldes commented:
/test aws-e2e-operator

jrvaldes commented:
/hold cancel

Release repo discussion is settled and https://github.com/openshift/release/pull/29791 merged.

jrvaldes commented:
/hold 

To investigate CI [issue](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_windows-machine-config-operator/1091/pull-ci-openshift-windows-machine-config-operator-master-aws-e2e-operator/1544704297508278272).

jrvaldes commented:
/test aws-e2e-upgrade

jrvaldes commented:
/test ci-index

jrvaldes commented:
/test aws-e2e-operator

jrvaldes commented:
/test aws-e2e-ccm-install

jrvaldes commented:
/test aws-e2e-ccm-install

jrvaldes commented:
/test aws-e2e-upgrade

jrvaldes commented:
/test aws-e2e-upgrade

jrvaldes commented:
/test aws-e2e-ccm-install

jrvaldes commented:
/test aws-e2e-upgrade

jrvaldes commented:
/test aws-e2e-ccm-install

jrvaldes commented:
/test aws-e2e-upgrade

jrvaldes commented:
/test aws-e2e-ccm-install

jrvaldes commented:
Depends on:
- https://github.com/ovn-org/ovn-kubernetes/pull/3074

jrvaldes commented:
/test aws-e2e-upgrade

jrvaldes commented:
> Depends on:
> 
> * [[hybrid-overlay] Forward routes for link-local addresses to vNIC in Windows ovn-org/ovn-kubernetes#3074](https://github.com/ovn-org/ovn-kubernetes/pull/3074)

Upstream and downstream sync PR, both merged:

- https://github.com/openshift/ovn-kubernetes/pull/1214

jrvaldes commented:
Depends on submodule update PR:

- https://github.com/openshift/windows-machine-config-operator/pull/1127

mansikulkarni96 commented:
@jrvaldes #1127 has merged this one can be opened for CI Cycles.

jrvaldes commented:
> @jrvaldes #1127 has merged this one can be opened for CI Cycles.

Correct, rebased, and updated the PR. Triggering AWS CI jobs.

/test aws-e2e-upgrade

jrvaldes commented:
/hold cancel

jrvaldes commented:
/test aws-e2e-operator

jrvaldes commented:
/test aws-e2e-ccm-install

jrvaldes commented:
/test aws-e2e-upgrade

jrvaldes commented:
/test aws-e2e-upgrade

jrvaldes commented:
Rebased with:
- https://github.com/openshift/windows-machine-config-operator/pull/1172

sebsoto commented on pkg/nodeconfig/nodeconfig.go:
```diff
@@ -467,15 +467,18 @@ func (nc *nodeConfig) Deconfigure() error {
-		patchData, meta.PatchOptions{})
-	if err != nil {
-		return errors.Wrapf(err, "error removing version annotation from node %s", nc.node.GetName())
+	// Check version annotation and remove iff present
```
How does this commit relate to adding Server 2022 support?

jrvaldes replied:
Removed.

sebsoto commented on pkg/instance/instance.go:
```diff
@@ -58,14 +58,8 @@ func (i *Info) UpgradeRequired() bool {
-	}
-
-	// Version annotation has an incorrect value, this was configured by an older version of WMCO and should be
+	// Version annotation is not present or has an incorrect value, means this instance
```
How does this commit relate to adding Server 2022 support?

jrvaldes replied:
As per our discussion, we are going to focus on the issue with the EC2 metadata endpoint not being available. Removed this commit as it was proposed as a walk-around.

jrvaldes commented:
/hold

To investigate the issue with AWS EC2 metadata endpoint not available.

jrvaldes commented:
/test aws-e2e-upgrade

jrvaldes commented:
/test aws-e2e-operator

jrvaldes commented:
/test aws-e2e-operator

aravindhp commented:
/close

This hack does not work. For one EC2Launch v2 is a service and we should be restarting that and not just executing the binary. Further more restarting the service does not result in the routes being recreated as the services loops waiting for an IP address:
```log
2023-02-03 00:33:31 Console: Network interface configuration is empty...waiting for an IP...
2023-02-03 00:34:35 Console: Network interface configuration is empty...waiting for an IP...
```
